### PR TITLE
feat(treesitter): support URLs

### PR DESCRIPTION
**Pull request by gpanders**
Tree-sitter queries can add URLs to a capture using the `#set!` directive, e.g.

    (inline_link
      (link_text) @text.reference
      (link_destination) @text.uri
      (#set! @text.reference "url" @text.uri))

The `#set!` directive is modified to accept other captures as the argument, which includes the captured node in the metadata table.

In order to support directives which reference other nodes, the highlighter must be updated to use `iter_matches` rather than `iter_captures` since the former provides the `match` table which maps capture IDs to nodes.

---

Blocked by:

- [x] https://github.com/neovim/neovim/pull/27131
- [x] https://github.com/neovim/neovim/pull/27109
- [X] https://github.com/neovim/neovim/pull/24738

### CODE REVIEW

**Comment by Danielkonge** [comment]
If the treesitter highlighter uses `iter_matches` instead of `iter_captures`, doesn't it run into the problem of #17099? I helped rewrite `rainbow-delimiters` to use `iter_captures` instead of `iter_matches`, because we couldn't get the highlighting we wanted with `iter_matches` when there were multiple captures with the same name in one match.

**Comment by gpanders** [comment]
@Danielkonge Yep you are absolutely right. So this PR is now blocked on _three_ other PRs.

File: runtime/lua/vim/treesitter/highlighter.lua [hunk 1]
```diff
@@ -259,19 +259,21 @@ local function on_line_impl(self, buf, line, is_spell_nav)
 
     if state.iter == nil or state.next_row < line then
       state.iter =
-        state.highlighter_query:query():iter_captures(root_node, self.bufnr, line, root_end_row + 1)
+        state.highlighter_query:query():iter_matches(root_node, self.bufnr, line, root_end_row + 1)
```
Comment (MariaSolOs) [inline runtime/lua/vim/treesitter/highlighter.lua:262]: Are we sure `state.highlighter_query:query()` won't be nil?

**Comment by gpanders** [comment]
In discussion on Matrix @wookayin brought up a very good point: can we use the `match` table in `iter_captures` and avoid moving to `iter_matches`?

It turns out, we can! `iter_captures` already includes the `match` table which maps capture ids to nodes, which is exactly what we need to support directives that use captures as values (as `#set!` does to set URLs). All we have to do is return it from the `iter_captures` iterator.

This is a much simpler change and makes the path forward for this PR clear (and we can probably also revert https://github.com/neovim/neovim/pull/27131 now, since it is no longer needed).

**Comment by gpanders** [comment]
With https://github.com/neovim/neovim/pull/24738 merged, I changed this back to use `iter_matches` for highlighting. All tests are passing!

@lewis6991 @clason @wookayin Please re-review when able. Let's ship this thing.

File: runtime/queries/markdown_inline/highlights.scm [hunk 2]
```diff
@@ -33,6 +33,11 @@
   ] @markup.link
   (#set! conceal ""))
 
+(inline_link
+  (link_text) @markup.link.label
+  (link_destination) @markup.link
+  (#set! @markup.link.label "url" @markup.link))
```
Comment (clason) [inline runtime/queries/markdown_inline/highlights.scm:39]: How does this interact with the conceal later? (What happens if `conceallevel` is set?)

**Comment by clason** [comment]
We probably want to add this to the `(uri_autolink)` node in `markdown_inline` and the `(url)` node in `vimdoc` as well.

(Adding support to this in nvim-treesitter will be a bigger effort...)

File: runtime/queries/markdown_inline/highlights.scm (hunk 2 above)
Comment (gpanders) [reply runtime/queries/markdown_inline/highlights.scm:39]: It works beautifully https://github.com/neovim/neovim/pull/27109#issuecomment-1902244611

**Comment by gpanders** [comment]
> We probably want to add this to the `(uri_autolink)` node in `markdown_inline` and the `(url)` node in `vimdoc` as well.
> 
> (Adding support to this in nvim-treesitter will be a bigger effort...)

It is not needed for nodes which are _actual_ URLS (like `uri_autolink`) because the terminal emulator will already linkify those (at least every terminal emulator I have ever seen that supports OSC 8 already auto linkifies URLs).

**Review by clason — APPROVED** [review]

File: runtime/lua/vim/treesitter/highlighter.lua [hunk 3]
```diff
@@ -258,51 +258,69 @@ local function on_line_impl(self, buf, line, is_spell_nav)
...
+          if hl and end_row >= line and (not is_spell_nav or spell ~= nil) then
+            local priority = (tonumber(metadata.priority) or vim.highlight.priorities.treesitter)
+              + spell_pri_offset
+            local url = metadata[capture] and metadata[capture].url ---@type string|number|nil
+            if type(url) == 'number' then
+              if match and match[url] then
+                -- Assume there is only one matching node. If there are more than one, take the URL
+                -- from the first.
+                local other_node = match[url][1]
+                url = vim.treesitter.get_node_text(other_node, buf, metadata and metadata[url])
```
Comment (wookayin) [inline runtime/lua/vim/treesitter/highlighter.lua:300]: Q: Is this line covered by tests? `get_node_text()` expects the third argument to be `opt = { metadata = ... }`, so it looks like we are not passing the correct argument.

**Review by wookayin — CHANGES_REQUESTED** [review]
I find there is a regression: injected highlighting doesn't work for me (e.g. lua -> comment). Also, conceal looks weird to me (e.g. vimdoc/markdown).

**Comment by gpanders** [comment]
> I find there is a regression: injected highlighting doesn't work for me (e.g. lua -> comment). Also, conceal looks weird to me (e.g. vimdoc/markdown).

I can repro the conceal breakage in vimdoc. We have a test for conceal, but apparently it's not extensive enough because it didn't catch this.

Can you give me a repro for the injected highlighting?

File: runtime/lua/vim/treesitter/highlighter.lua (hunk 3 above)
Comment (gpanders) [reply runtime/lua/vim/treesitter/highlighter.lua:300]: Alas, it was not that easy. I suspect the problem lies deeper (another bug in `iter_matches`).

Comment (gpanders) [reply runtime/lua/vim/treesitter/highlighter.lua:300]: Ok I think I understand what's happening. The vimdoc query uses patterns like this:

```
(tag
  "*" @markup.heading.5.marker
  (#set! conceal "")
  text: (_) @label)
```

But with `iter_matches`, the `#set!` directive now applies to the *whole pattern*, whereas with `iter_captures` it would have applied only to the previous capture(?).

Comment (clason) [reply runtime/lua/vim/treesitter/highlighter.lua:300]: Hmm, that's a fairly common pattern in nvim-treesitter (also with predicates).

Comment (gpanders) [reply runtime/lua/vim/treesitter/highlighter.lua:300]: The solution is to make it more explicit about _what_ should be concealed:

```
(taglink
  "|" @markup.link.delimiter
  (#set! @markup.link.delimiter conceal "")
  text: (_) @markup.link)
```

This is (imo) more explicit and easier to understand. But at the obvious cost of being (yet another) breaking change.

Comment (clason) [reply runtime/lua/vim/treesitter/highlighter.lua:300]: I consider that acceptable (since indeed better because more explicit). And fairly straightforward to fix, though tedious.

Comment (clason) [reply runtime/lua/vim/treesitter/highlighter.lua:300]: For highlight queries; adapting injections would be a mess...

Comment (clason) [reply runtime/lua/vim/treesitter/highlighter.lua:300]: (Probably also applies to priority.)

Comment (wookayin) [reply runtime/lua/vim/treesitter/highlighter.lua]: Please never mind about Markdown. I think conceals are fine with Markdown. For injections I will try to find a repro and let you know in another place.

Using the new vimdoc highlight query, conceals in vimdoc now look fine. One concern is that this new query doesn't work fully backward-compatible with the previous highlighter implementation (nvim 0.9.x or on the current master): "`", `*`, `|` are no longer hidden. This might be a bit problematic because the query files from `nvim-treesitter` also need to be updated.

Comment (gpanders) [reply runtime/lua/vim/treesitter/highlighter.lua]: I trust @clason to handle the situation with nvim-treesitter. When this PR is merged we can update the queries on the `main` branch of nvim-treesitter which is intended to work with 0.10 only.

**Comment by wookayin** [comment]
Here is a repro for highlights with language injection:

```lua
for name, url in pairs {
  ['nvim-treesitter'] = 'https://github.com/nvim-treesitter/nvim-treesitter',
} do
  local install_path = vim.fn.fnamemodify('nvim_issue/' .. name, ':p')
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
  -- Important: nvim-treesitter must precede $VIMRUNTIME in &runtimepath
  vim.opt.runtimepath:prepend(install_path)
end

-- ensure lua and luadoc parser (override the bundled one because of injection queris)
require('nvim-treesitter.install').ensure_installed({ 'lua', 'luadoc' })

vim.cmd [[
  hi! link @comment.luadoc            @comment.documentation
  hi!      @keyword.luadoc            guifg=#a488a6 gui=NONE
  hi! link @keyword.return.luadoc     @keyword.luadoc    " @return
]]
```

image

Expected behavior: `@param`, `@return`, etc. in luadoc should be highlighted. `:Inspect` (using `iter_captures`) can detect them.

File: runtime/lua/vim/treesitter/highlighter.lua (hunk 3 above)
Comment (clason) [reply runtime/lua/vim/treesitter/highlighter.lua]: Might have to bite the bullet and backport (part of) this.

**Comment by gpanders** [comment]
Thanks @wookayin, I can repro.

My guess is that this is related to the ordering that extmarks are applied, which means there are still some cases that our `subpriority` fix is not catching.

**Comment by gpanders** [comment]
I've added a failing test case for language injections. Now just need to fix the code to make the test case pass.

File: test/functional/treesitter/highlight_spec.lua [hunk 4]
```diff
@@ -697,7 +703,7 @@ describe('treesitter highlighting (C)', function()
                                                                        |
         LuaRef cb = nlua_ref(, 1);                                     |
                                                                        |
-        multiqueue_put(main_loop.events, nlua_schedule_event,          |
+        {11:V}(main_loop.events, nlua_schedule_event,                       |
```
Comment (lewis6991) [inline test/functional/treesitter/highlight_spec.lua:706]: What is this voodoo?

**Review by lewis6991 — APPROVED** [review]

**Comment by gpanders** [comment]
The issue with injections is that injections depend on the order in which the extmarks are applied to be visible over the "base" highlights. Consider this test Lua file:

```lua
cdef([[int x = 5;]]);
```

When the highlighter for the Lua query runs, it highlights `[[int x = 5;]]` as a string. The pattern index for this in the Lua highlight query is 45, so we use `_subpriority = 45`. When the highlighter runs again for the C highlight query, it highlights `int` with `@type.builtin`. The pattern index for this in the C highlight query is 36, so we use `_subpriority = 36`. Because the pattern index is lower than the one for the Lua string, the injection is not visible.

When we don't use subpriorities this isn't an issue because the extmark for the injection is applied after the base highlight. But we _must_ use subpriorities when using `iter_matches` precisely because we _cannot_ rely on the order that extmarks are applied.

My solution for this is to track the largest pattern index seen in each tree, and offset the pattern index in subsequent trees by that amount. When we do this, then the `@string` highlight from the Lua query uses subpriority `45` and the `@type.builtin` highlight form the C query uses subpriority `45 + 36 = 81`.

File: runtime/lua/vim/treesitter/highlighter.lua [hunk 5]
```diff
@@ -248,6 +248,13 @@ end
 ---@param line integer
 ---@param is_spell_nav boolean
 local function on_line_impl(self, buf, line, is_spell_nav)
+  -- Track the maximum pattern index encountered in each tree. For subsequent
+  -- trees, the subpriority passed to nvim_buf_set_extmark is offset by the
+  -- largest pattern index from the prior tree. This ensures that extmarks
+  -- from subsequent trees always appear "on top of" extmarks from previous
+  -- trees (e.g. injections should always appear over base highlights).
```
Comment (lewis6991) [inline runtime/lua/vim/treesitter/highlighter.lua:255]: This specifically isn't working for me. Which part of the code aims to ensure injection highlights takes priority?

Comment (lewis6991) [reply runtime/lua/vim/treesitter/highlighter.lua:255]: Ok, I've debugged the issue. The problem is `on_line_impl` can (and often does) highlight more than 1 line.

My example:

```bash
echo "
module dut (input clk, reset_n);
endmodule
" > dut.sv
```

Here the contents of `echo` are an injection. However the base tree will highlight the string argument to `echo` in the call to `on_line_imp(line=0)`.

When `on_line_impl(line=1)` is called, there is no highlight state in the base tree and so when the injection is highlighted, `pattern_offset=0`.

This causes the highlight from the base tree to have a higher subpriority than the injection.

So basically I don't think this method is going to work. What might work is offsetting the priority based on how many parents the langtree has.

Comment (lewis6991) [reply runtime/lua/vim/treesitter/highlighter.lua:255]: https://github.com/neovim/neovim/pull/27844 should fix it.
